src/bt-pan-io.c

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright (C) 2015  Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+#include <fdio/task.h>
+#include <hardware/bluetooth.h>
+#include <hardware/bt_pan.h>
+#include <stdlib.h>
+#include "compiler.h"
+#include "log.h"
+#include "bt-proto.h"
+#include "bt-pdubuf.h"
+#include "bt-core-io.h"
+#include "bt-pan-io.h"
+
+enum {
+  /* commands/responses */
+  OPCODE_ENABLE = 0x01,
+  OPCODE_GET_LOCAL_ROLE = 0x02,
+  OPCODE_CONNECT = 0x03,
+  OPCODE_DISCONNECT = 0x04,
+  /* notifications */
+  OPCODE_CONTROL_STATE_NTF = 0x81,
+  OPCODE_CONNECTION_STATE_NTF = 0x82,
+};
+
+static void (*send_pdu)(struct pdu_wbuf* wbuf);
+static const btpan_interface_t* btpan_interface;
+
+static enum ioresult
+send_ntf_pdu(void* data)
+{
+  /* send notification on I/O thread */
+  if (!send_pdu) {
+    ALOGE("send_pdu is NULL");
+    return IO_OK;
+  }
+  send_pdu(data);
+  return IO_OK;
+}
+
+static enum ioresult
+send_cmd_pdu(void* data)
+{
+  /* send command on I/O thread */
+  if (!send_pdu) {
+    ALOGE("send_pdu is NULL");
+    return IO_OK;
+  }
+  send_pdu(data);
+  return IO_OK;
+}
+
+/*
+ * Notifications
+ */
+
+static void
+control_state_cb(btpan_control_state_t state, bt_status_t error,
+                 int local_role, const char* ifname)
+{
+  struct pdu_wbuf* wbuf;
+
+  wbuf = create_pdu_wbuf(1 + /* control state */
+                         1 + /* status */
+                         1 + /* local role */
+                         17, /* interface name */
+                         0, NULL);
+  if (!wbuf)
+    return;
+
+  init_pdu(&wbuf->buf.pdu, SERVICE_BT_PAN, OPCODE_CONTROL_STATE_NTF);
+  if ((append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)state) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)error) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)local_role) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "0", ifname) < 0))
+    goto cleanup;
+
+  //padding
+  void* padding;
+  int len;
+  len = 17 - strnlen(ifname, 17);
+  if(len > 0){
+    padding = malloc(len);
```

Comment:
Some additional words about |malloc| and friends. It has quite a bit of overhead: from a few 100 cycles to whatever page allocation in the kernel takes. So please don't call it unless you really need it. In this code, you'd leak the memory and you'd write the malloc'ed memory's content into the PDU; therefore leaking information.